### PR TITLE
Switch to Thinking::Sphinx real-time indexes

### LINKED ISSUE #8541: Searching for a recent created project doesn't return any match
# Issue Description
Search indexes seems not to be updated just after every object (project, package, ...) has been created. This results in no matches searching by their names, just after they have been created.

Now it is needed to wait for the reindexing proccess, that occurs each hour, for that project to be found in the search.

Expected Result
============
Searching for the project that has been just created, returns it as match, without waiting so much as for the next hourly reindex.

How to Reproduce
===============
1. Got to https://build.opensuse.org/
1. Create a project.
1. Search by the name of that project in the search box placed in the upper right corner.
1. See no matches.

Further Information
===============
* Thinking Sphinx: https://freelancing-gods.com/thinking-sphinx/
* Definition of our indexes: https://github.com/openSUSE/open-build-service/tree/master/src/api/app/indices/
* Enabled matches_regex for MySql by jamespearson · Pull Request #36800: https://github.com/rails/rails/pull/36800

### PR DESCRIPTION
**Pull request by eduardoj**
Instead of reindexing every hour with SQL-backed indexes, we now use real-time indices, that allow us to search for a just created or updated project or package.

Real-time indices are updated with callbacks, so we don't need an external process for updating indices.

Fixes #8547.

After this is fully tested we should also remove some remainig code, like the `FullTextIndexJob` job, the entry to reindex in the `config/clock.rb` file, the `obs-delayedjob-queue-sphinx-indexing.service` systemd service, ...

For a detailed explanation of what has been done, please read carefully the commits' descriptions.

Co-authored-by: David Kang <dkang@suse.com>
Co-authored-by: Saray Cabrera Padrón <scabrerapadron@suse.de>

### CODE REVIEW

**Comment by eduardoj** [comment]
Marking as "DO NOT MERGE" because both tests suites must be adapted.

File: src/api/Procfile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 web: bundle exec rails server -b 0.0.0.0
 delayed: bundle exec script/delayed_job.api.rb run
-clock: bundle exec clockworkd --log-dir=log -l -c config/clock.rb run 
-search: bundle exec rake ts:rebuild NODETACH=true
+clock: bundle exec clockworkd --log-dir=log -l -c config/clock.rb run
+search: (sleep 60 && bundle exec rake ts:rt:index) & bundle exec rake ts:rebuild NODETACH=true && fg
```
Comment (hennevogel) [inline src/api/Procfile]: `ts:rt:index` is not taken care of by `ts:rebuild`?

Comment (saraycp) [reply src/api/Procfile]: The normal sequence for `rebuild` is:
```
task :rebuild => [
    :stop, :clear, :configure, 'ts:sql:index', :start, 'ts:rt:index'
]
```

However, running just `ts:rebuild` causes errors at this point. Only if we add `NODETACH=true` it works. And adding that makes the sequence stop in `start`, it never runs  `ts:rt:index`.

**Comment by eduardoj** [comment]
We have updated the pull request, and now is ready to review. Please have a look at the updated descriptions of the pull request and commits.

To test the functionality you can use the review-app. Create a package or a project. After that you will find them in the results of a search.

Removing the "DO NOT MERGE" label.

File: src/api/app/models/package.rb [hunk 2]
```diff
@@ -1457,6 +1460,28 @@ def last_build_reason(repo, arch, package_name = nil)
     PackageBuildReason.new(data)
   end
 
+  # Methods for Thinking Sphinx real-time indexes
```
Comment (hennevogel) [inline src/api/app/models/package.rb]: Can you isolate them in some file please instead of working around the ClassLength cop.

File: src/api/app/models/package.rb [hunk 3]
```diff
@@ -1457,6 +1460,28 @@ def last_build_reason(repo, arch, package_name = nil)
...
+  def package_issues_ids
+    package_issues.pluck(:issue_id)
+  end
+
+  def devel_packages?
+    develpackages.exists?
+  end
+
+  def backend_packages?
+    Package.find_by_sql("SELECT * FROM backend_packages WHERE package_id = #{id}").present?
```
Comment (hennevogel) [inline src/api/app/models/package.rb]: This is a [cache](https://github.com/openSUSE/open-build-service/wiki/The-BackendPackage-Cache), you know this right?

File: src/api/app/models/project.rb [hunk 4]
```diff
@@ -1559,6 +1561,36 @@ def self.very_important_projects_with_categories
     end
   end
 
+  # Methods for Thinking Sphinx real-time indexes
```
Comment (hennevogel) [inline src/api/app/models/project.rb]: Same here...

File: src/api/spec/features/webui/search_spec.rb [hunk 5]
```diff
@@ -1,223 +1,226 @@
-require 'sphinx_helper'
+require 'browser_helper'
```
Comment (hennevogel) [inline src/api/spec/features/webui/search_spec.rb]: But why? It was nicely isolated in that helper. Now you have `thinking_sphinx_test_init` all over the place.

File: src/api/Procfile (hunk 1 above)
Comment (hennevogel) [reply src/api/Procfile]: :thought_balloon:  I really don't like this....

File: src/api/spec/features/webui/search_spec.rb (hunk 5 above)
Comment (eduardoj) [reply src/api/spec/features/webui/search_spec.rb]: They way RSpec helper methods are isolated is [documented](https://relishapp.com/rspec/rspec-core/docs/helper-methods/define-helper-methods-in-a-module#include-a-module-in-only-some-example-groups).

I have already checked, and calling `thinking_sphinx_test_init` in a test that doesn't have `thinking_sphinx: true` results in an `undefined local variable or method 'thinking_sphinx_test_init'` error.

In other words, `thinking_sphinx_test_init` is only available in tests with `thinking_sphinx: true`.

File: src/api/Procfile (hunk 1 above)
Comment (eduardoj) [reply src/api/Procfile]: I don't like it either. Any other alternative?

File: ReleaseNotes-2.11 [hunk 6]
```diff
@@ -29,6 +29,8 @@ Features
 Generic:
  * You can track javascript errors that occur in the users browser with errbit, enable
    it in the API configuration file options.yml.
+ * Search was improved: searching for a just created or updated project or package will
```
Comment (dmarcoux) [inline ReleaseNotes-2.11]: I suggest this:

`Search was improved: New and updated projects/packages are indexed directly instead of once every hour.`

Comment (eduardoj) [reply ReleaseNotes-2.11]: Thanks. Done.

File: src/api/app/models/package.rb (hunk 2 above)
Comment (eduardoj) [reply src/api/app/models/package.rb]: Done. Now they are isolated in `app/mixins/package_sphinx.rb`.

Comment (eduardoj) [reply src/api/app/models/package.rb]: However adding the two `include` calls made `app/models/package.rb` bigger, so I changed the `Max` limit of the `Metrics/ClassLength` cop.

File: src/api/Procfile (hunk 1 above)
Comment (eduardoj) [reply src/api/Procfile]: In production this `ts:rt:index` job is run after the `:start` job because `NODETACH` is NOT set to `true`. In production we are running it in detached mode.

File: src/api/app/models/package.rb (hunk 3 above)
Comment (eduardoj) [reply src/api/app/models/package.rb]: I changed this line to `backend_package.present?`, making use of the `has_one` relationship.

Being a cache, do you see it could be harmfull to use this query using real-time indices?

File: src/api/Procfile (hunk 1 above)
Comment (hennevogel) [reply src/api/Procfile]: Okay, then how about you create an `thinking_sphinx` issue and link it here?

File: src/api/app/models/package.rb (hunk 3 above)
Comment (hennevogel) [reply src/api/app/models/package.rb]: Well potentially, if the case doesn't exist during your `backend_package.present?` call, it will be [created](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/package.rb#L1107).

File: src/api/spec/features/webui/search_spec.rb (hunk 5 above)
Comment (eduardoj) [reply src/api/spec/features/webui/search_spec.rb]: The code to init and stop `ThinkingSphinx::Test` was moved from the tests to the `spec/support/thinking_sphinx.rb` file, using [hooks filtered by symbols](https://relishapp.com/rspec/rspec-core/v/3-9/docs/hooks/filters#filtering-hooks-using-symbols).

File: src/api/Procfile (hunk 1 above)
Comment (eduardoj) [reply src/api/Procfile]: Done: https://github.com/pat/thinking-sphinx/issues/1150

**Comment by coolo** [comment]
I don't think it's fixing #8547 - so what's the actual issue fixed? your branch is named fix_8541 - that might give a hint :)

**Comment by hennevogel** [comment]
#8541 was the issue fixed.

**Comment by saraycp** [comment]
I've just updated the description, we wrote the wrong number.

**Comment by coolo** [comment]
Note that with this a running searchd is now a requirement. If you create/update projects and searchd doesn't run, you get a 500 now - it wasn't a problem before. Struggled with that in the CI of openSUSE-release-tools

**Comment by hennevogel** [comment]
> If you create/update projects and searchd doesn't run, you get a 500 now - it wasn't a problem before. 

This was always a problem.

**Comment by coolo** [comment]
It was most likely a problem to find the projects afterwards through the webui, but for our CI there was no requirement on searchd in the API before.
